### PR TITLE
Trigger payloads

### PR DESCRIPTION
**Pull request by robfletcher**
The mistake I made with the previous implementation of trigger types was thinking that the `type` field specified the kind of payload the trigger contains. This is wrong, as a manual trigger can have Jenkins build info, for example.

This implementation instead identifies the type based on the fields that exist in the JSON. This means any type of trigger can contain any type of payload.

### CODE REVIEW

**Comment by robfletcher** [comment]
@lwander I have a failing test here to do with artifacts. Any chance you could take a look? I'm not sure what I've broken.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Trigger.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  val type: String,
+  @get:JsonUnwrapped val payload: TriggerPayload? = null,
+  val correlationId: String? = null,
+  val user: String? = "[anonymous]",
+  val parameters: Map<String, Any> = mutableMapOf(),
+  val artifacts: List<Artifact> = mutableListOf(),
+  val notifications: List<Map<String, Any>> = mutableListOf(),
+  var isRebake: Boolean = false,
+  var isDryRun: Boolean = false,
+  var isStrategy: Boolean = false
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Trigger.kt]: All triggers are this type. Variance is in the `payload` field.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/TriggerPayload.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.model
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonProperty
+
+interface TriggerPayload
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/TriggerPayload.kt]: Base for different types of payload

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/TriggerPayload.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    @param:JsonProperty("relativePath") val relativePath: String
+  )
+}
+
+data class DockerTriggerPayload
+@JsonCreator constructor(
+  @JsonProperty("account") val account: String,
+  @JsonProperty("repository") val repository: String,
+  @JsonProperty("tag") val tag: String?
+) : TriggerPayload
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/TriggerPayload.kt]: example of a payload type.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+  private fun JsonNode.parseTriggerPayload(parser: JsonParser): TriggerPayload? {
+    return when {
+      looksLikeDocker() -> parseValue<DockerTriggerPayload>(parser)
+      looksLikeJenkins() -> parseValue<JenkinsTriggerPayload>(parser)
+      looksLikePipeline() -> parseValue<PipelineTriggerPayload>(parser)
+      looksLikeGit() -> parseValue<GitTriggerPayload>(parser)
+      else -> null
+    }
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt]: Identifies what type of payload based on the fields that exist in the incoming JSON.

Comment (lwander) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt]: What if it's a webhook with an arbitrary payload, does this just become null?

Comment (robfletcher) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt]: How would anything in Orca reference it?

Comment (lwander) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt]: Using SPEL

Comment (robfletcher) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt]: ah, yeah. Maybe I should add a catch all here that's just a `Map<String, Object>`

**Comment by robfletcher** [comment]
I've added an `UnknownTriggerPayload` class that can just contain anything.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/DefaultTrigger.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.model
+
+import com.netflix.spinnaker.kork.artifacts.model.Artifact
+import com.netflix.spinnaker.kork.artifacts.model.ExpectedArtifact
+
+data class DefaultTrigger
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/DefaultTrigger.kt]: Common trigger interface

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/DockerTrigger.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  override val user: String? = "[anonymous]",
+  override val parameters: Map<String, Any> = mutableMapOf(),
+  override val artifacts: List<Artifact> = mutableListOf(),
+  override val notifications: List<Map<String, Any>> = mutableListOf(),
+  override var isRebake: Boolean = false,
+  override var isDryRun: Boolean = false,
+  override var isStrategy: Boolean = false,
+  val account: String,
+  val repository: String,
+  val tag: String?
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/DockerTrigger.kt]: additional fields for docker

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/GitTrigger.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  override val parameters: Map<String, Any> = mutableMapOf(),
+  override val artifacts: List<Artifact> = mutableListOf(),
+  override val notifications: List<Map<String, Any>> = mutableListOf(),
+  override var isRebake: Boolean = false,
+  override var isDryRun: Boolean = false,
+  override var isStrategy: Boolean = false,
+  val source: String,
+  val project: String,
+  val branch: String,
+  val slug: String
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/GitTrigger.kt]: additional fields for Git

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/JenkinsTrigger.kt [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  override val parameters: Map<String, Any> = mutableMapOf(),
+  override val artifacts: List<Artifact> = mutableListOf(),
+  override val notifications: List<Map<String, Any>> = mutableListOf(),
+  override var isRebake: Boolean = false,
+  override var isDryRun: Boolean = false,
+  override var isStrategy: Boolean = false,
+  val master: String,
+  val job: String,
+  val buildNumber: Int,
+  val propertyFile: String?
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/JenkinsTrigger.kt]: Additional fields for Jenkins

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/JenkinsTrigger.kt [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+) : Trigger {
+
+  override var other: Map<String, Any> = mutableMapOf()
+  override var resolvedExpectedArtifacts: List<ExpectedArtifact> = mutableListOf()
+  var buildInfo: BuildInfo? = null
+  var properties: Map<String, Any> = mutableMapOf()
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/JenkinsTrigger.kt]: Properties and build info are assigned later when looked up from Igor

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/PipelineTrigger.kt [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  override val correlationId: String? = null,
+  override val user: String? = "[anonymous]",
+  override val parameters: Map<String, Any> = mutableMapOf(),
+  override val artifacts: List<Artifact> = mutableListOf(),
+  override val notifications: List<Map<String, Any>> = mutableListOf(),
+  override var isRebake: Boolean = false,
+  override var isDryRun: Boolean = false,
+  override var isStrategy: Boolean = false,
+  val parentExecution: Execution,
+  val parentPipelineStageId: String? = null
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/PipelineTrigger.kt]: additional fields for pipeline triggers

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Trigger.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  val parameters: Map<String, Any>
+  val artifacts: List<Artifact>
+  val notifications: List<Map<String, Any>>
+  var isRebake: Boolean
+  var isDryRun: Boolean
+  var isStrategy: Boolean
+  var resolvedExpectedArtifacts: List<ExpectedArtifact>
+  @set:JsonAnySetter
+  @get:JsonAnyGetter
+  var other: Map<String, Any>
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Trigger.kt]: Common trigger fields available on any type. The `other` field will capture any free-form data.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    hasNonNull("account") && hasNonNull("repository") && hasNonNull("tag")
+
+  private fun JsonNode.looksLikeGit() =
+    hasNonNull("source") && hasNonNull("project") && hasNonNull("branch") && hasNonNull("slug")
+
+  private fun JsonNode.looksLikeJenkins() =
+    hasNonNull("master") && hasNonNull("job") && hasNonNull("buildNumber")
+
+  private fun JsonNode.looksLikePipeline() =
+    hasNonNull("parentExecution")
```
Comment (robfletcher) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt]: Above methods determine type by the fields present in the JSON, ignoring the `type` field.

**Pull request merged**